### PR TITLE
Remove cyclic dependencies

### PR DESCRIPTION
**Pull request by knazarov**
This is required to embed luarocks into tarantool, as current tarantool preloader can't preload cyclic dependencies.
There should be a unidirectional dependency graph and predictable order.

Note: as a consequence of this PR, operating systems other that unix-compatible ones are no longer supported. This is because I had to manually resolve dependency graph for predictable require() order.

### CODE REVIEW

File: src/luarocks/fs/lua.lua [hunk 1]
```diff
@@ -870,4 +870,16 @@ function fs_lua.is_lua(name)
...
+   for name, fn in pairs(fs_table) do
+      if not fs[name] then
+         fs[name] = fn
+      end
+   end
+end
+
+-- Load platform-independent pure-Lua functionality
+load_fns(fs_lua)
+
```
Comment (Totktonada) [inline src/luarocks/fs/lua.lua:883]: Double blank line.

File: src/luarocks/fs/tools.lua [hunk 2]
```diff
@@ -153,4 +153,14 @@ function tools.get_md5(file)
...
+local function load_fns(fs_table)
+   for name, fn in pairs(fs_table) do
+      if not fs[name] then
+         fs[name] = fn
+      end
+   end
+end
+load_fns(tools)
+
+
```
Comment (Totktonada) [inline src/luarocks/fs/tools.lua:165]: Double blank line.

File: src/luarocks/fs/tools.lua [hunk 3]
```diff
@@ -153,4 +153,14 @@ function tools.get_md5(file)
...
 end
 
+local function load_fns(fs_table)
+   for name, fn in pairs(fs_table) do
+      if not fs[name] then
+         fs[name] = fn
+      end
+   end
+end
+load_fns(tools)
```
Comment (Totktonada) [inline src/luarocks/fs/tools.lua]: No blank line between a function and a statement.

File: src/luarocks/fs/unix.lua [hunk 4]
```diff
@@ -132,4 +132,14 @@ function unix.tmpname()
...
 end
 
+local function load_fns(fs_table)
+   for name, fn in pairs(fs_table) do
+      if not fs[name] then
+         fs[name] = fn
+      end
+   end
+end
+load_fns(unix)
```
Comment (Totktonada) [inline src/luarocks/fs/unix.lua]: No blank line between a function and a statement.

File: src/luarocks/fs/unix.lua [hunk 5]
```diff
@@ -132,4 +132,14 @@ function unix.tmpname()
...
+
```
Comment (Totktonada) [inline src/luarocks/fs/unix.lua:142]: Double blank line.

File: src/luarocks/fs/unix/tools.lua [hunk 6]
```diff
@@ -234,4 +234,14 @@ function tools.make_temp_dir(name)
...
 end
 
+local function load_fns(fs_table)
+   for name, fn in pairs(fs_table) do
+      if not fs[name] then
+         fs[name] = fn
+      end
+   end
+end
+load_fns(tools)
```
Comment (Totktonada) [inline src/luarocks/fs/unix/tools.lua:244]: No blank line between a function and a statement.

File: src/luarocks/fs/unix/tools.lua [hunk 7]
```diff
@@ -234,4 +234,14 @@ function tools.make_temp_dir(name)
...
+
```
Comment (Totktonada) [inline src/luarocks/fs/unix/tools.lua:245]: Double blank line.

File: src/luarocks/fs.lua [hunk 8]
```diff
@@ -42,35 +42,4 @@ fs.verbose = function()    -- patch io.popen and os.execute to display commands
...
-   for name, fn in pairs(fs_table) do
-      if not fs[name] then
-         fs[name] = fn
-      end
-   end
-end
-
--- Load platform-specific functions
-local loaded_platform = nil
-for _, platform in ipairs(cfg.platforms) do
```
Comment (Totktonada) [inline src/luarocks/fs.lua:55]: Now we ignore config's OS autodetection. Commit message should explicitly state that, I think, because it is change of the behaviour.

File: src/luarocks/command_line.lua [hunk 9]
```diff
@@ -10,6 +10,17 @@ local path = require("luarocks.path")
...
 local deps = require("luarocks.deps")
 local fs = require("luarocks.fs")
+--- This is to install platform-specific fs functions.
+-- Since there were cyclic dependencies between fs.* modules
+-- and it was very hard to compile luarocks into Tarantool
+-- with them (an artifact of current preloader implementation),
+-- they were broken down. As a consequence, we have to manually
+-- load the platform-specific fs functions here, or they
+-- just won't work.
+require("luarocks.fs.unix")
```
Comment (Totktonada) [inline src/luarocks/command_line.lua:20]: Now we don't pcall it, so unix is mandatory. Tarantool doesn't support other OSes, so it can be considered ok. But I don't like the idea mix non-so-related changes as drop win32 support and prevent cycling in require. It is not a blocker, I just think that we can prevent cyclic deps w/o these changes and Occam's razor push me dive into this nitpicking. Feel free to ignore this comment except proper commenting in the commit message any changes of the behaviour.

File: src/luarocks/loader.lua [hunk 10]
```diff
@@ -22,27 +22,9 @@ local manif_core = require("luarocks.manif_core")
 local deps = require("luarocks.deps")
 local util = require("luarocks.util")
 
--- Workaround for wrappers produced by older versions of LuaRocks
```
Comment (Totktonada) [inline src/luarocks/loader.lua:25]: I think such things should be described in the commit message and possibly even be moved into separate commit.

File: src/luarocks/manif.lua [hunk 11]
```diff
@@ -344,6 +343,8 @@ local function store_results(results, manifest)
    assert(type(results) == "table")
    assert(type(manifest) == "table")
 
+   local repos = require('luarocks.repos')
```
Comment (Totktonada) [inline src/luarocks/manif.lua:349]: I would comment it like 'prevent cyclic dependencies' just to give a reader an idea that this has the certain reason.

File: src/luarocks/manif_core.lua [hunk 12]
```diff
@@ -49,6 +48,7 @@ function manif_core.manifest_loader(file, repo_url, lua_version, quick)
    end
    local globals = err
    if not quick then
+      local type_check = require('luarocks.type_check')
```
Comment (Totktonada) [inline src/luarocks/manif_core.lua:54]: Again: I would comment it like 'prevent cyclic dependencies'.

File: src/luarocks/search.lua [hunk 13]
```diff
@@ -176,7 +175,9 @@ function search.manifest_search(results, repo, query, lua_version)
    assert(type(results) == "table")
    assert(type(repo) == "string")
    assert(type(query) == "table")
-   
+
+   local manif = require('luarocks.manif')
```
Comment (Totktonada) [inline src/luarocks/search.lua:182]: Again: I would comment it like 'prevent cyclic dependencies'.

**Review by Totktonada — APPROVED** [review]
It looks good in general, but some style comments are below. I'm kinda afraid to drop support of something (especially w/o unavoidable reason), but give you LGTM, because don't know how it can broke something.

@Gerold103, can you please look into this PR?

**Review by Gerold103 — APPROVED** [review]
I do not know anything about luarocks and do not use it. So can not comprehend the patch nor leave an in-depth review.

File: src/luarocks/command_line.lua (hunk 9 above)
Comment (knazarov) [reply src/luarocks/command_line.lua:20]: Squashed review fixes and reworded commit message to mention this.
